### DISCUSSION TITLE
How to use CableConstraint achieve rotation?

### DISCUSSION THREAD

**Discussion by TeoREN0217** [SofaDefrost/SoftRobots#194] [Help - Programming with SoftRobots] [answered]
Hi! I'm trying to using only one cable to control a continuum robot. After add CableConstraint to the robot, I want to do a extra rotation ( or twist) besides displacements, which means I need to rotate the Cable by an axis. 
I was trying three methods all not worked:

1. Rotate the robot model instead of rotate the cable. But it didn't for the reason I thought that we build mapping between robot model and cable model. So rotate the robot model will rotate the cable at the same time.
2. I also tried to build a rotation controller like `WholeGripperController` in tutorials `PneunetGripper`, using the function `rotateRestPos(rest_pos,rx,centerPosY,centerPosZ)` to change the `MechanicalObject` of my cable directly. In the function it changed the `rest_position` and I have successfully pass the value in, but there's no excepted twisting motion.
3. Passing the `Transformation` of   `MechanicalObject` which has `rotation` and 'rotation2', both not worked.

So I am wondering is there any methods to achieve the function that cable rotate and make the robot model twist? And I found that `reset_position` and `rest_postion`; also the `Transformation` of   `MechanicalObject` has `rotation` and `rotation2`. What's the difference   between them？ Looking forward to your reply~

**Comment by EulalieCoevoet** [comment]
Hi, 

It makes me think of [this paper](https://www.youtube.com/watch?time_continue=87&v=qwzKAgw31pU&embeds_euri=https%3A%2F%2Fteam.inria.fr%2F&source_ve_path=MTM5MTE3&feature=emb_logo). Maybe @adagolodjo can point you examples?

**Reply by TeoREN0217** [reply]
Hi @EulalieCoevoet ! Thanks so much for your reply. I have read the paper and I think that's what I want to do. Looking for @adagolodjo 's reply~

**Comment by TeoREN0217** [comment, marked as answer]
Hi! @EulalieCoevoet and @adagolodjo

I have successfully achieved teh function according to the paper link you mentioned. For others who may have the same problem as me, here is the method:
Using `directFingerActuator` template in `Cosserat Plugin` and I think there are some bugs: 

In function `def onKeypressedEvent(self, event) ` ` self._extracted_from_onKeypressedEvent_10( qOld, posA, self.angularRate)` has an extra param `self`, you need to delete it. 
And in my SOFA version, `TetrahedronSetTopologyContainer` didn't work well. So I changed it as `MeshTopology`.

Hope these can help others!

**Reply by EulalieCoevoet** [reply]
Thanks for your feedback @TeoREN0217! I'm glad it worked for you.

**Comment by adagolodjo** [comment]
Hello @TeoREN0217 ,

Thanks for your feedback, and sorry for my late reply. 
I am happy that you can use the article to solve your problem.
For the error in the scene, this is because the scene is based on an older version of SOFA and SofaPython3. 
Feel free to create an issue (that, I can look at when I have a free moment) or a pull-request. Your contribution is necessary for the update and the evolution of the plugin. 
Thanks, @EulalieCoevoet for the notification.
